### ISSUE TITLE
Support for more data sources: MS Sql Server, Sybase

### ISSUE THREAD

**Issue by turtl7** [grafana/grafana#1692]
Some of our projects are having databases with MS Sql Server, Sybase. 
Can Grafana support these as data sources?
[I see list of available GO drivers at https://github.com/golang/go/wiki/SQLDrivers]
Thanks.

**Comment by torkelo** [comment]
MS SQL could be added in the future, I want to have an official windows build shortly, maybe for 2.1 or 2.2

**Comment by gsaray101** [comment]
:+1:

**Comment by perseo-dev** [comment]
I have tried to implement extensions for mssql in a separate local branch but without success; i catch always nil pointer or invalid memory address; maybe because I'm using SQL Azure, or more probably because I'm a .NET Developer and I don't know golang; it's a pity that you've decided to implement the backend in golang; I think that Node+JS would allow more people to contribute. anyway you've made an excellent work! thank you

**Comment by torkelo** [comment]
@perseo-dev that is strange. I am actually coming from .NET (worked as C# dev for 10 years), now only work in linux though. But I want to have a good windows build and metrics experiance for Grafana in the future. Not sure what dialect SQL Azure is (regular MSSQL?). 

I choose Go because I want to be able to do fast numerical transformations on the backend in the future (node is not good for fast float operations). And the language is very hot right now within the server software & devops scene which helps a lot with many contributions.

**Comment by perseo-dev** [comment]
Can you tell me how debug the code in windows with free tools? Thanks

--- Messaggio originale ---

Da: "Torkel Ödegaard" <notifications@github.com>
Inviata: 25 Maggio 2015 09:49
A: "grafana/grafana" <grafana@noreply.github.com>
Cc: "perseo-dev" <perseo-dev@outlook.com>
Oggetto: Re: [grafana] Support for more data sources: MS Sql Server, Sybase (#1692)

@perseo-dev that is strange. I am actually coming from .NET (worked as C# dev for 10 years), now only work in linux though. But I want to have a good windows build and metrics experiance for Grafana in the future. Not sure what dialect SQL Azure is (regular MSSQL?).

I choose Go because I want to be able to do fast numerical transformations on the backend in the future (node is not good for fast float operations). And the language is very hot right now within the server software & devops scene which helps a lot with many contributions.

---

**Comment by turtl7** [comment]
I saw Windows specific build available. Any luck  getting Sql Server supported sooner? 
Appreciated for giving a try.

**Comment by agilgur5** [comment]
@torkelo what types of transformations are you planning? Go on the back end and Angular on the front end with little comments and documentation (on the code, not how to use the UI) in most places really make the learning curve steep. Node + React would've made contributions way easier, and a separate Go process for these numerical transformations could've been added later; you don't need the web server itself to be in Go, and I haven't found an API layer better than Node + ActionHero. Agreed that Go is becoming more popular in devops, but Node is far, far more popular for web apps in general

**Comment by torkelo** [comment]
@agilgur5 Grafana is not a library, I have almost no time to document the UI and configuration, you spect code guides and docs as well? Angular is a lot more popular than React. Although I would love to have something else than Angular, I hate angular, but would take half a year full time to rewrite it. 

I love nodejs and it would have been a good choice for the backend, but it has some big performance issues around float handing that the grafana backend requires.

**Comment by torkelo** [comment]
@agilgur5 btw, I focus a lot on code readability and unit tests to ease contributions, seems to be working. Contributions to the backend has been plentiful :)

**Comment by agilgur5** [comment]
@torkelo just so you know, I'm not trying to bash you or anything, there's a lot of features I'd love to contribute so I can use this more, after using Grafana for a week or so now, but I found it hard to figure out what to do based on past code in part due to unfamiliarity with Go, so I'm just questioning your decisions and suggesting alternatives for the future.

I'd say Angular is only more popular because of its age, not its usability (even you hate it!) :P I'd definitely consider forking and starting to migrate some code over to React once generic SQL support is ready.

The way I do most code documentation is auto-generated just based on my comments (usually via Docco), which is why I believe adding docs shouldn't take too much time in general (not just for this project). Could you elaborate on the float handling that Grafana requires? Since it's not in the main codebase right now, I would've thought to just have a Go microservice to do any float handling numerical transforms, but run Node as the main server process. 

Anyway, those are my opinions, sorry for slightly derailing this thread! Thanks for all the work torkelo, I know the pains of being a one-man team :)

**Comment by torkelo** [comment]
Grafana will never migrate to React I think, its just too much code :) Maybe though. It depends on how angular 2 shapes up. 

But I would never do SQL from react :) not sure what made you mention "generic SQL" in context of React?

One of the main reasons of having Go is to have a single binary without any runtime dependencies. Makes it very easy to distribute and install. I would love to have a more micro service architecture (I am a big proponent of it, and used it in the past), but it would require so much work for install guides, install packages, distribution, build packages & config. Having single install binary and & install experiance is crucial for open source projects.

**Comment by agilgur5** [comment]
Yea I have no idea how angular 2 will end up (though I doubt better than React :P). And oh no, by generic SQL I just meant that the milestone including generic SQL is my biggest use case, so once that milestone is achieved, I would like to contribute to this more (like with migrating to React, for instance).

Ahhh, yes you're completely right about the single binary part. This is why I'm a fan of Docker; write once, deploy anywhere securely. How I've been doing micro services with Docker is to have a single install script that builds multiple containers and then runs and links them together -- so all anyone needs is one install script. You could also put all of those containers into another container (Docker in Docker) to simplify it further. Then it can run on any OS (you can't install Grafana on Mac directly yet), no need for separate packages for different distributions, same install experience everywhere. Obviously that has the potential of running into more bugs than a single install binary though.

Thanks for the clarification! That makes a lot more sense now :)

**Comment by nickman** [comment]
@torkelo;
It seems you will never be done handling requests to implement support for X, Y and Z. It occurred to me that it might be helpful to invert the integration pattern. This is what made me thing of it:  Grafana has top-tier support for Graphite, at least, it seems that way to me. I'm an OpenTSDB user and once in a while, I get envious of those self populating drop downs and magically appearing annotations. I know these items are works in progress, but I am impatient, so I started looking at the code that makes the template drop downs populate from a Graphite query. The problem is, I am a javascript idiot and half of it is complete greek to me (apologies to any Greeks). But then I thought, why does it have to Graphite at the other end of the socket ? The Graphite query API is pretty simple, and I can put pretty much whatever I want into the query field. All I need is a listener at the other end that is pretending to be Graphite, accepts whatever metric name symbolics I choose to model my protocol on, and fullfill the query with a graphite compliant response from *whatever data source* I choose. (yeah, to me that's easier than writing 100 lines of javascript....)

So the idea of inverting the pattern would be that Grafana publishes a standard API that datasources need to implement. Let's safely assume it's JSON over HTTP or even WebSockets because they're better for callbacks (did someone ask about InfluxDB continuous queries ?). I suppose you could even use the Graphite protocol straight out or as the basis, but it could be more generalized to support a wider variety of integrations. So now Grafana is configured with the endpoints any particular user wants to use and it calls out to them using a standard API. Now the nodejs types (who would probably find such an API trivial to implement) can whip up all the services they want in nodejs. The Windows guys can run their implementation on Windows, they can use nodejs, C#, java or whatever they choose, provided it can listen on HTTP/WebSocket requests and provide compliant responses. They can get their data from SQLServer or (<insert 127 more databases here>). I know a guy who would use CA Spectrum as an annotation source. (he might be the only person in the galaxy that would, but he knows it and would never seriously open a Github ticket asking you to implement it....) And I'm not suggesting "instead-of" the built in services, but rather "in addition to".

What do you think ?

**Comment by eosfor** [comment]
Hi, any good news about MS SQL as a source? Actually why i'm asking. The simplest way to put perf counters to the scene from numerous windows boxes is to configure data collectors and put data directly to ms sql without any "middle ware collectors". So it would be really great to have an option jut to read that data from MS SQL. So this would be a really great option!

**Comment by Alcon835** [comment]
I know this one hasn't been commented on in awhile, but this is a feature I know my and other Enterprise environments would be very interested in. I'm aware that alerting is the big focus now, but once that has launched, is this on the list?

**Comment by ptMuta** [comment]
@Alcon835 you can checkout the current progress at #5364

**Comment by philip-wernersbach** [comment]
The proxy that I talk about in #5364 would work with MS SQL and Sybase with a little modification, or you could look into contributing the code to #5364.

**Comment by Alcon835** [comment]
Thank you!

**Comment by philip-wernersbach** [comment]
@Alcon835 @all I've open sourced my InfluxDB-to-MySQL protocol converter and published it to https://github.com/philip-wernersbach/influx-mysql, and it's ready to work with Grafana.

**Comment by styfle** [comment]
@torkelo SQL Server 2017 is no longer exclusive to Windows--[Linux](https://www.microsoft.com/en-us/sql-server/sql-server-2017) is supported 🎉 

In light of that information and the MySQL PR landing, is there any chance to get MS SQL Server data source support in Grafana? 😄

**Comment by qgicup** [comment]
Really need as well the MS SQL Server data source support in Grafana. Willing to drop PowerBi if that works fine in Grafana.

**Comment by dmitrios** [comment]
I had to create custom datasource that consumes web API ( running on .NET core) to execute queries on SQL Server. I do think that using common API format for SQL scenarios would be the best way to quickly add support for multiple types of databases.

**Comment by bkez322** [comment]
I would benefit a lot from MS SQL as well.

**Comment by mohanraj1939** [comment]
Hi,
For MSSQL there is a seperate plugin available in github. Follow the link given below.

[https://github.com/gbrian/grafana-simple-sql-datasource](url)

**Comment by linuxchips** [comment]
Hi, I stumbled upon this issue now. I have made a pull request #10093 for the mssql support a while ago. I have originally made the changes to v4.6 and now I have merged my fork to v5
I have only tested it as a data source, I have currently 4 dashboards using it. I have tested the annotations but not with a real data, just played around with it.
the data source supports multi column values (where the metric name is the column name) and native datetime types.
and I also added the ability to have a metric column along with the multi column value, the metric name in that case will be "metricname - columnname"
feel free to try it and give me some feed back
